File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/Indicators/Indicator.java
package com.github.anastr.speedviewlib.components.Indicators;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;

import com.github.anastr.speedviewlib.Speedometer;

/**
 * this Library build By Anas Altair
 * see it on <a href="https://github.com/anastr/SpeedView">GitHub</a>
 */
@SuppressWarnings("unchecked,unused,WeakerAccess")
public abstract class Indicator<I extends Indicator> {

    protected Paint indicatorPaint =  new Paint(Paint.ANTI_ALIAS_FLAG);
    private float density;
    private float indicatorWidth;
    private float viewSize;
    private float speedometerWidth;
    private int indicatorColor = 0xff2196F3;
    private int padding;
    private boolean inEditMode;

    protected Indicator (Context context) {
        this.density = context.getResources().getDisplayMetrics().density;
        init();
    }

    private void init() {
        indicatorPaint.setColor(indicatorColor);
        indicatorWidth = getDefaultIndicatorWidth();
    }

    public abstract void draw(Canvas canvas, float degree);
    /** called when size change or color, width */
    protected abstract void updateIndicator();
    /** @param withEffects if indicator have effects like BlurMaskFilter */
    protected abstract void setWithEffects(boolean withEffects);
    protected abstract float getDefaultIndicatorWidth();

    /**  @return Y position of indicator */
    public float getTop(){
        return getPadding();
    }
    /** @return Bottom Y position of indicator */
    public float getBottom(){
        return getCenterY();
    }
    /** @return down point after center */
    public float getLightBottom() {
        return getCenterY() > getBottom() ? getBottom() : getCenterY();
    }

    /**
     * must call in {@code speedometer.onSizeChanged()}
     * @param speedometer target speedometer.
     */
    public void onSizeChange(Speedometer speedometer) {
        setTargetSpeedometer(speedometer);
    }

    /**
     * to change indicator's data,
     * this method called by the library.
     * @param speedometer target speedometer.
     */
    public void setTargetSpeedometer(Speedometer speedometer) {
        updateData(speedometer);
        updateIndicator();
    }

    private void updateData(Speedometer speedometer) {
        this.viewSize = speedometer.getSize();
        this.speedometerWidth = speedometer.getSpeedometerWidth();
        this.padding = speedometer.getPadding();
        this.inEditMode = speedometer.isInEditMode();
    }

    public float dpTOpx(float dp) {
        return dp * density;
    }

    public float getIndicatorWidth() {
        return indicatorWidth;
    }

    public I setIndicatorWidth(float indicatorWidth) {
        this.indicatorWidth = indicatorWidth;
        return (I) this;
    }

    /**
     * @return size of Speedometer View without padding.
     */
    public float getViewSize() {
        return viewSize - (padding*2f);
    }

    public int getIndicatorColor() {
        return indicatorColor;
    }

    public I setIndicatorColor(int indicatorColor) {
        this.indicatorColor = indicatorColor;
        return (I) this;
    }

    /**
     * @return x center of speedometer.
     */
    public float getCenterX() {
        return viewSize /2f;
    }

    /**
     * @return y center of speedometer.
     */
    public float getCenterY() {
        return viewSize /2f;
    }

    public int getPadding() {
        return padding;
    }

    public float getSpeedometerWidth() {
        return speedometerWidth;
    }

    public void noticeIndicatorWidthChange(float indicatorWidth) {
        this.indicatorWidth = indicatorWidth;
        updateIndicator();
    }

    public void noticeIndicatorColorChange(int indicatorColor) {
        this.indicatorColor = indicatorColor;
        updateIndicator();
    }

    public void noticeSpeedometerWidthChange(float speedometerWidth) {
        this.speedometerWidth = speedometerWidth;
        updateIndicator();
    }

    public void noticePaddingChange(int newPadding) {
        this.padding = newPadding;
        updateIndicator();
    }

    public void withEffects(boolean withEffects) {
        setWithEffects(withEffects);
        updateIndicator();
    }

    public boolean isInEditMode() {
        return inEditMode;
    }

    /** indicator's shape */
    public enum Indicators {
        NoIndicator, NormalIndicator, NormalSmallIndicator, TriangleIndicator
        , SpindleIndicator, LineIndicator, HalfLineIndicator, QuarterLineIndicator
        , KiteIndicator, NeedleIndicator
    }

    /**
     * create new {@link Indicator} with default values.
     * @param context required.
     * @param indicator new indicator (Enum value).
     * @return new indicator object.
     */
    public static Indicator createIndicator (Context context, Indicators indicator) {
        switch (indicator) {
            case NoIndicator :
                return new NoIndicator(context);
            case NormalIndicator :
                return new NormalIndicator(context);
            case NormalSmallIndicator :
                return new NormalSmallIndicator(context);
            case TriangleIndicator :
                return new TriangleIndicator(context);
            case SpindleIndicator :
                return new SpindleIndicator(context);
            case LineIndicator :
                return new LineIndicator(context, LineIndicator.LINE);
            case HalfLineIndicator :
                return new LineIndicator(context, LineIndicator.HALF_LINE);
            case QuarterLineIndicator :
                return new LineIndicator(context, LineIndicator.QUARTER_LINE);
            case KiteIndicator :
                return new KiteIndicator(context);
            case NeedleIndicator :
                return new NeedleIndicator(context);
            default :
                return new NormalIndicator(context);
        }
    }
}

File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/Indicators/Indicator.kt
package com.github.anastr.speedviewlib.components.Indicators

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint

import com.github.anastr.speedviewlib.Speedometer

/**
 * this Library build By Anas Altair
 * see it on [GitHub](https://github.com/anastr/SpeedView)
 */
abstract class Indicator<I : Indicator<I>> protected constructor(context: Context) {

    protected var indicatorPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val density: Float
    private var indicatorWidth: Float = 0.toFloat()
    private var viewSize: Float = 0.toFloat()
    var speedometerWidth: Float = 0.toFloat()
        private set
    private var indicatorColor = -0xde690d
    var padding: Int = 0
        private set
    var isInEditMode: Boolean = false
        private set
    protected abstract val defaultIndicatorWidth: Float

    /**  @return Y position of indicator
     */
    open fun getTop(): Float = padding.toFloat()
    /** @return Bottom Y position of indicator
     */
    open fun getBottom(): Float = getCenterY()
    /** @return down point after center
     */
    fun getLightBottom(): Float = if (getCenterY() > getBottom()) getBottom() else getCenterY()

    /**
     * @return x center of speedometer.
     */
    fun getCenterX(): Float = viewSize / 2f

    /**
     * @return y center of speedometer.
     */
    fun getCenterY(): Float = viewSize / 2f

    init {
        this.density = context.resources.displayMetrics.density
        init()
    }

    private fun init() {
        indicatorPaint.color = indicatorColor
        indicatorWidth = defaultIndicatorWidth
    }

    abstract fun draw(canvas: Canvas, degree: Float)
    /** called when size change or color, width  */
    protected abstract fun updateIndicator()

    /** @param withEffects if indicator have effects like BlurMaskFilter
     */
    protected abstract fun setWithEffects(withEffects: Boolean)

    /**
     * must call in `speedometer.onSizeChanged()`
     * @param speedometer target speedometer.
     */
    fun onSizeChange(speedometer: Speedometer) {
        setTargetSpeedometer(speedometer)
    }

    /**
     * to change indicator's data,
     * this method called by the library.
     * @param speedometer target speedometer.
     */
    fun setTargetSpeedometer(speedometer: Speedometer) {
        updateData(speedometer)
        updateIndicator()
    }

    private fun updateData(speedometer: Speedometer) {
        this.viewSize = speedometer.size.toFloat()
        this.speedometerWidth = speedometer.getSpeedometerWidth()
        this.padding = speedometer.padding
        this.isInEditMode = speedometer.isInEditMode
    }

    fun dpTOpx(dp: Float): Float {
        return dp * density
    }

    fun getIndicatorWidth(): Float {
        return indicatorWidth
    }

    fun setIndicatorWidth(indicatorWidth: Float): I {
        this.indicatorWidth = indicatorWidth
        return this as I
    }

    /**
     * @return size of Speedometer View without padding.
     */
    fun getViewSize(): Float {
        return viewSize - padding * 2f
    }

    fun getIndicatorColor(): Int {
        return indicatorColor
    }

    fun setIndicatorColor(indicatorColor: Int): I {
        this.indicatorColor = indicatorColor
        return this as I
    }

    fun noticeIndicatorWidthChange(indicatorWidth: Float) {
        this.indicatorWidth = indicatorWidth
        updateIndicator()
    }

    fun noticeIndicatorColorChange(indicatorColor: Int) {
        this.indicatorColor = indicatorColor
        updateIndicator()
    }

    fun noticeSpeedometerWidthChange(speedometerWidth: Float) {
        this.speedometerWidth = speedometerWidth
        updateIndicator()
    }

    fun noticePaddingChange(newPadding: Int) {
        this.padding = newPadding
        updateIndicator()
    }

    fun withEffects(withEffects: Boolean) {
        setWithEffects(withEffects)
        updateIndicator()
    }

    /** indicator's shape  */
    enum class Indicators {
        NoIndicator, NormalIndicator, NormalSmallIndicator, TriangleIndicator, SpindleIndicator, LineIndicator, HalfLineIndicator, QuarterLineIndicator, KiteIndicator, NeedleIndicator
    }

    companion object {

        /**
         * create new [Indicator] with default values.
         * @param context required.
         * @param indicator new indicator (Enum value).
         * @return new indicator object.
         */
        fun createIndicator(context: Context, indicator: Indicators): Indicator<*> {
            when (indicator) {
                Indicator.Indicators.NoIndicator -> return NoIndicator(context)
                Indicator.Indicators.NormalIndicator -> return NormalIndicator(context)
                Indicator.Indicators.NormalSmallIndicator -> return NormalSmallIndicator(context)
                Indicator.Indicators.TriangleIndicator -> return TriangleIndicator(context)
                Indicator.Indicators.SpindleIndicator -> return SpindleIndicator(context)
                Indicator.Indicators.LineIndicator -> return LineIndicator(context, LineIndicator.LINE)
                Indicator.Indicators.HalfLineIndicator -> return LineIndicator(context, LineIndicator.HALF_LINE)
                Indicator.Indicators.QuarterLineIndicator -> return LineIndicator(context, LineIndicator.QUARTER_LINE)
                Indicator.Indicators.KiteIndicator -> return KiteIndicator(context)
                Indicator.Indicators.NeedleIndicator -> return NeedleIndicator(context)
                else -> return NormalIndicator(context)
            }
        }
    }
}
